### PR TITLE
SLUDGE: Added support for impulse tracker files using libmikmod!

### PR DESCRIPTION
**Pull request by hari01584**
Changes
1. Added libmikmod compilation and scripts to configure!
2. Impulse tracker mod wrappers in audio/mods!
3. Testbed engine updated to support .it files playback!
4. Sludge engine sound updated to add support for impulse tracker files!

### CODE REVIEW

File: audio/mods/impulsetracker.cpp [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+ImpulseTrackerMod::ImpulseTrackerMod(Common::SeekableReadStream *stream, DisposeAfterUse::Flag disposeAfterUse) {
+	if (!stream) {
+		warning("ImpulseTrackerMod::ImpulseTrackerMod(): Input file/stream is invalid.");
+		return;
+	}
+
+	MikMod_InitThreads();
+
+	// No sound driver as we will be routing through scrumm mixer!
```
Comment (sev-) [inline audio/mods/impulsetracker.cpp:160]: What is "scrumm mixer"?

File: audio/mods/impulsetracker.cpp [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+		warning("ImpulseTrackerMod::ImpulseTrackerMod(): Could not initialize sound, reason: %s\n",
+		        MikMod_strerror(MikMod_errno));
+		return;
+	}
+
+	// Loading only impulse tracker loader!
+	MikMod_RegisterLoader(&load_it);
+
+
+	// TODO: Not currently in use
```
Comment (sev-) [inline audio/mods/impulsetracker.cpp:176]: Is it truly not in use now?

File: audio/mods/impulsetracker.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+	_stream = stream;
+
+	_dispose = disposeAfterUse;
+
+	// Load mod using custom loader class!
+	_reader = createMikMemoryReader(_stream);
+	_mod = Player_LoadGeneric(_reader, 64, 0);
+	if (!_mod) {
+		warning("ImpulseTrackerMod::ImpulseTrackerMod(): Parsing mod error : %s", MikMod_strerror(MikMod_errno));
```
Comment (sev-) [inline audio/mods/impulsetracker.cpp:185]: Please remove space before colon `error: %s`

File: audio/mods/impulsetracker.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/**
+ * @file
+ * Sound decoder used in engines:
+ */
+#define FORBIDDEN_SYMBOL_EXCEPTION_FILE
```
Comment (sev-) [inline audio/mods/impulsetracker.h:26]: Why do you need it here?

File: audio/module.mk [hunk 5]
```diff
@@ -47,6 +47,7 @@ MODULE_OBJS := \
 	mods/maxtrax.o \
 	mods/mod_xm_s3m.o \
 	mods/module.o \
+	mods/impulsetracker.o \
```
Comment (sev-) [inline audio/module.mk:50]: This must be sorted alphabetically

File: configure [hunk 6]
```diff
@@ -5497,6 +5508,35 @@ fi
...
+#include <mikmod.h>
+int main(void) {
+	SAMPLE *spl = Sample_Load("abcd");
+	Sample_Free(spl);
+	return 0;
+}
+EOF
+	if test -n "$_host"; then
+		# don't execute while cross compiling
+		cc_check $MIKMOD_CFLAGS $MIKMOD_LIBS -lmikmod && _mikmod=yes
```
Comment (sev-) [inline configure:5527]: Please use ` libmikmod-config` to obtain proper compilation and linker flags. See `freetype-config` in this file for a good example.

File: engines/testbed/sound.cpp [hunk 7]
```diff
@@ -354,4 +361,4 @@ SoundSubsystemTestSuite::SoundSubsystemTestSuite() {
 	addTest("SampleRates", &SoundSubsystem::sampleRates, true);
 }
 
-} // End of namespace Testbed
+} // End of namespace Testbed
```
Comment (sev-) [inline engines/testbed/sound.cpp:364]: Why did you remove end line here?

**Review by sev- — CHANGES_REQUESTED** [review]
I did my review

File: audio/mods/impulsetracker.cpp (hunk 1 above)
Comment (hari01584) [reply audio/mods/impulsetracker.cpp:160]: Yea, it like meant going through the inbuilt system to play, rather than using mikmod internal players, i think its better to just remove this comment

File: audio/mods/impulsetracker.cpp (hunk 2 above)
Comment (hari01584) [reply audio/mods/impulsetracker.cpp:176]: It is, however wrapped in MREADER instance, this is like for now we just remove this class parameter __stream_ and still the program will work

**Comment by criezy** [comment]
sev already did a thorough review.

I have one additional comment though: the split of the changes between the two commits is a mess, and the first commit likely does not compile. This will need to be fixed by moving some changes between the commits.

**Comment by hari01584** [comment]
That is indeed correct, I was thinking of squashing these two commits into one, will that be okay?

**Comment by hari01584** [comment]
Agreed, please check the latest changes, split into 3 commits each for engine/task!

**Comment by sev-** [comment]
All the changes look OK now, except for this new commit log message.

Please, fix it to "AUDIO: Add support for Impulse Tracker using libmikmod"

**Comment by sev-** [comment]
Thanks!
